include/arch/arm/aarch32/misc.h

```diff
@@ -31,10 +32,35 @@ static ALWAYS_INLINE void arch_nop(void)
 	__asm__ volatile("nop");
 }
 
+#if defined(CONFIG_USERSPACE)
+#if defined(CONFIG_CPU_CORTEX_M)
+#include <arch/arm/aarch32/cortex_m/cmsis.h>
+#elif defined(CONFIG_CPU_CORTEX_R)
+#include <arch/arm/aarch32/cortex_a_r/cmsis.h>
+#endif
+
+static inline bool z_arm_thread_is_in_user_mode(void)
```

Comment:
I think we need to move this function in the private API.
This file includes public , arch_ APIs, @bbolen

Reply:
Since include/arch/arm/aarch32/syscall.h arch_is_user_context() uses z_arm_thread_is_in_user_mode(), I assume that can't work since the syscall public header would need to reference a private header.  Am I misunderstanding?

Reply:
@ioannisg If we move this to a private header under zephyr/arch/arm/include, then we have to add ${ARCH_DIR}/${ARCH}/include to the top level cmake file so that anyone who can possibly include syscall.h also gets the private header that defines z_arm_thread_is_in_user_mode.  For example, kernel/userspace.c would need the include path for zephyr/arch/arm/include because it includes syscall.h.  I checked and no other architecture includes private headers in the public API.  I think the best options are
1)  I can put syscall.h back to the way it was and open code arch_is_user_context with the contents of z_arm_thread_is_in_user_mode.
2) I can put the prototype for z_arm_thread_is_in_user_mode in the public api, and put the definition down in zephyr/arch/arm.  This removes the inline ability of the function.  There are several instances of people already doing this, e.g. z_arm_irq_direct_dynamic_dispatch_reschedule, z_arm_irq_priority_set, z_arm_save_fp_context.

Reply:
@ioannisg ping

Reply:
Guess option (2) is the safest.